src/full/Agda/TypeChecking/Rules/Application.hs

```diff
@@ -1044,23 +1044,47 @@ inferOrCheckProjApp e o ds args mt = do
         [ "  principal arg " <+> prettyTCM arg
         , "  has type "      <+> prettyTCM ta
         ]
-      inferOrCheckProjAppToKnownPrincipalArg e o ds args mt k v0 ta
+      inferOrCheckProjAppToKnownPrincipalArg e o ds args mt k v0 ta Nothing
 
 -- | Same arguments 'inferOrCheckProjApp' above but also gets the position,
 --   value and type of the principal argument.
-inferOrCheckProjAppToKnownPrincipalArg ::
-  A.Expr -> ProjOrigin -> List1 QName -> A.Args -> Maybe (Comparison, Type) ->
-  Int -> Term -> Type -> TCM (Term, Type, CheckedTarget)
-inferOrCheckProjAppToKnownPrincipalArg e o ds args mt k v0 ta = do
+inferOrCheckProjAppToKnownPrincipalArg
+  :: A.Expr
+     -- ^ The whole expression which constitutes the application.
+  -> ProjOrigin
+     -- ^ The origin of the projection involved in this projection application.
+  -> List1 QName
+     -- ^ The projection name (potentially ambiguous).
+  -> A.Args
+     -- ^ The arguments to the projection.
+  -> Maybe (Comparison, Type)
+     -- ^ The expected type of the expression (if 'Nothing', infer it).
+  -> Int
+     -- ^ The position of the principal argument.
+  -> Term
+     -- ^ The value of the principal argument.
+  -> Type
+     -- ^ The type of the principal argument.
+  -> Maybe PrincipalArgTypeMetas
+     -- ^ The metas previously created for the principal argument's type, when
+     --   picking up a postponed problem. 'Nothing', otherwise.
+  -> TCM (Term, Type, CheckedTarget)
+     -- ^ The type-checked expression and its type (if successful).
+inferOrCheckProjAppToKnownPrincipalArg e o ds args mt k v0 ta mpatm = do
   let cmp = caseMaybe mt CmpEq fst
-      postpone b = do
+      postpone b vargs ta = do
         tc <- caseMaybe mt newTypeMeta_ (return . snd)
-        v <- postponeTypeCheckingProblem (CheckProjAppToKnownPrincipalArg cmp e o ds args tc k v0 ta) b
+        v <- postponeTypeCheckingProblem (CheckProjAppToKnownPrincipalArg cmp e o ds args tc k v0 ta (PrincipalArgTypeMetas vargs ta)) b
         return (v, tc, NotCheckedTarget)
   -- ta should be a record type (after introducing the hidden args in v0)
-  (vargs, ta) <- implicitArgs (-1) (not . visible) ta
+  (vargs, ta) <- caseMaybe mpatm
```

Comment:
Here you could rather construct a `PrincipalArgTypeMetas` pair:
```haskell
PrincipalArgTypeMetas vargs ta <- case mpatm of
  Just patm -> return patm
  Nothing -> uncurry PrincipalArgTypeMetas <$> implicitArgs (-1) (not . visitble) ta
```

Reply:
Oooooh, clever! Thanks for pointing this out. What I did seemed a little convoluted even to me, but I couldn't come up with anything better. Good to learn about `uncurry`. Makes sense that there would be a dual to `curry`. I'll make that change, too.

Reply:
Arg, sorry, I forgot something... :$  (I feel embarrased.)
```haskell
patm@(PrincipleArgTypeMetas vargs ta) <- case..
...
...  (postpone ... patm)
```
(This is the point of the whole story of not constructing a `(,)`-tuple.)

Reply:
Teehee... Yes. That makes sense.

No worries. I'm happy to keep iterating. We're still getting strictly better with each iteration!

I rebased onto current `master`, made the change, and it seems like things still work.

Have a good weekend!

Reply:
Thanks for your patience.  I finally merged it. :)